web/server/web_client.c

```diff
@@ -812,109 +849,348 @@ static inline char *http_header_parse(struct web_client *w, char *s, int parse_u
     }
     else if(!strncmp(s, "STREAM ", 7)) {
 #ifdef ENABLE_HTTPS
-        if ( (w->ssl.flags) && (netdata_use_ssl_on_stream & NETDATA_SSL_FORCE)){
-            w->header_parse_tries = 0;
-            w->header_parse_last_size = 0;
-            web_client_disable_wait_receive(w);
+        //CASE THESE CONDITIONS ARE TRUE, THE CLIENT IS SENDING A MESSAGE THAT IS NOT
+        //ENCRYPTED, SO I MUST LOG AND RETURN AN ERROR FOR IT.
+        if ((w->ssl.flags) && (netdata_use_ssl_on_stream & NETDATA_SSL_FORCE)) {
             char hostname[256];
             char *copyme = strstr(s,"hostname=");
-            if ( copyme ){
+            if (copyme) {
                 copyme += 9;
                 char *end = strchr(copyme,'&');
-                if(end){
+                if(end) {
                     size_t length = end - copyme;
                     memcpy(hostname,copyme,length);
                     hostname[length] = 0X00;
-                }
-                else{
+                } else {
                     memcpy(hostname,"not available",13);
                     hostname[13] = 0x00;
                 }
-            }
-            else{
+            } else {
                 memcpy(hostname,"not available",13);
                 hostname[13] = 0x00;
             }
+
             error("The server is configured to always use encrypt connection, please enable the SSL on slave with hostname '%s'.",hostname);
-            return HTTP_VALIDATION_NOT_SUPPORTED;
+            return NULL;
         }
 #endif
-
-        encoded_url = s = &s[7];
+        s = &s[7];
         w->mode = WEB_CLIENT_MODE_STREAM;
+    } else {
+        s = NULL;
     }
-    else {
+
+    return s;
+}
+
+/**
+ *  Find Protocol
+ *
+ *  The end of the url is the place where starts the protocol of the message,
+ *  this function finds the exact position where it starts.
+ *
+ * @param w is the structure with all information of the client request.
+ * @param s is the start of the user request
+ *
+ * @return
+ */
+static inline char *web_client_find_protocol(struct web_client *w,char *s) {
+    s = url_find_protocol(s);
+    if (s) {
+        //I must add 1, because I searched the separator+protocol
+        w->protocol.body = s+1;
+        //I know this string has at least ' HTTP/', so I do not need to search from the beginning
+        char *end = strchr(s+6,'\n');
+        if (end) {
+            w->protocol.length = end - w->protocol.body ;
+        }
+    }
+
+    return s;
+}
+
+/**
+ *  Parse header
+ *
+ *  Parse the headers sent together with the URL.
+ *
+ * @param w is the structure with all information of the client request.
+ * @param s is the string where the URL finished
+ */
+static inline void web_client_parse_headers(struct web_client *w,char *s) {
+    while(*s) {
+        // find a line feed
+        while(*s && *s++ != '\r');
+
+        // did we reach the end?
+        if(unlikely(!*s)) break;
+
+        if (*s == '\n') {
```

Comment:
Logic: what if there is no `'\n'`, is this a problem?

Reply:
This function I am carrying from the previous parser, it is not a problem this part of the code, because this is the header parser(The headers are from the second line until the last line of the request), and we already saw in the previous functions of the algorithm that we will have at least two '\n' at the end of the request.